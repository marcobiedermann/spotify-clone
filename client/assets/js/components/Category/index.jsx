import PropTypes from 'prop-types';
import React from 'react';
import { Link } from 'react-router-dom';
import './style.css';

const Category = ({ id, icons, name }) => (
  <figure className="category">
    {icons[0] && (
      <Link to={`/browse/categories/${id}/playlists`}>
        <img src={icons[0].url} alt={name} />
      </Link>
    )}
    <figcaption>
      <Link to={`/browse/categories/${id}/playlists`}>
        {name}
      </Link>
    </figcaption>
  </figure>
);

Category.propTypes = {
  id: PropTypes.string,
  icons: PropTypes.arrayOf(PropTypes.shape()),
  name: PropTypes.string,
};

Category.defaultProps = {
  id: '',
  icons: [],
  name: '',
};

export default Category;
