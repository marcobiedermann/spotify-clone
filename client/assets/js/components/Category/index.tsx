import React, { FC } from 'react';
import { Link } from 'react-router-dom';
import Image, { ImageProps } from '../Image';

export interface CategoryProps {
  icons: ImageProps[];
  id: string;
  name: string;
}

const Category: FC<CategoryProps> = (props) => {
  const { id, icons, name } = props;

  return (
    <figure>
      {icons[0] && (
        <Link to={`/browse/categories/${id}/playlists`}>
          <Image {...icons[0]} alt={name} />
        </Link>
      )}
      <figcaption>
        <Link to={`/browse/categories/${id}/playlists`}>{name}</Link>
      </figcaption>
    </figure>
  );
};

export default Category;
