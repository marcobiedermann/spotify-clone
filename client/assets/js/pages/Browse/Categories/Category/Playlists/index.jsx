import PropTypes from 'prop-types';
import React, { Component } from 'react';
import {
  Route,
  Switch,
} from 'react-router-dom';
import Playlists from '../../../../../components/Playlists';
import { ACCESS_TOKEN } from '../../../../../constants/config';

class PlaylistsPage extends Component {
  componentDidMount() {
    const { fetchCategoryPlaylists } = this.props;

    fetchCategoryPlaylists(ACCESS_TOKEN, 'dinner');
  }

  render() {
    const { match, playlists } = this.props;

    return (
      <Switch>
        <Route
          path={`${match.url}`}
          component={() => <Playlists playlists={playlists.items} />}
        />
      </Switch>
    );
  }
}

PlaylistsPage.propTypes = {
  playlists: PropTypes.shape({
    items: PropTypes.arrayOf(PropTypes.shape),
  }),
  fetchCategoryPlaylists: PropTypes.func,
  match: PropTypes.shape({
    url: PropTypes.string,
  }),
};

PlaylistsPage.defaultProps = {
  playlists: {
    items: [],
  },
  fetchCategoryPlaylists: () => {},
  match: {
    url: '',
  },
};

export default PlaylistsPage;
