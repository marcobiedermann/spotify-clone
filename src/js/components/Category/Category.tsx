import { Link } from 'react-router-dom';
import Image from '../Image';

const defaultIconSize = 100;

interface Icon {
  height: number | null;
  url: string;
  width: number | null;
}

export interface CategoryProps {
  icons: Icon[];
  id: string;
  name: string;
}

function Category(props: CategoryProps): JSX.Element {
  const { id, icons, name } = props;
  const icon = icons[0];

  return (
    <figure>
      {icons[0] && (
        <Link to={`/browse/categories/${id}/playlists`}>
          <Image
            {...icon}
            alt={name}
            width={icon.width || defaultIconSize}
            height={icon.height || defaultIconSize}
          />
        </Link>
      )}
      <figcaption>
        <Link to={`/browse/categories/${id}/playlists`}>{name}</Link>
      </figcaption>
    </figure>
  );
}

export default Category;
