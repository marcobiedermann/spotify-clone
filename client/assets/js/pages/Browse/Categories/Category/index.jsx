import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Helmet } from 'react-helmet';
import { connect } from 'react-redux';
import { Route, Switch } from 'react-router-dom';
import { bindActionCreators } from 'redux';
import PlaylistsPage from './Playlists';
import { fetchCategory } from '../../../../actions/browse';
import Category from '../../../../components/Category';
import Error from '../../../../components/Error';
import Loader from '../../../../components/Loader';

export class CategoryPage extends Component {
  static propTypes = {
    accessToken: PropTypes.string,
    category: PropTypes.shape({
      name: PropTypes.string,
    }),
    error: PropTypes.shape({
      message: PropTypes.string,
    }),
    fetchCategory: PropTypes.func,
    isLoading: PropTypes.bool,
    match: PropTypes.shape({
      params: PropTypes.shape({
        category_id: PropTypes.string,
      }),
      url: PropTypes.string,
    }),
  };

  static defaultProps = {
    accessToken: '',
    category: {
      name: '',
    },
    error: {
      message: '',
    },
    fetchCategory: () => {},
    isLoading: false,
    match: {
      params: {
        category_id: '',
      },
      url: '',
    },
  };

  componentDidMount() {
    const { accessToken, fetchCategory, match } = this.props;

    fetchCategory(accessToken, match.params.category_id);
  }

  render() {
    const {
      category, error, isLoading, match,
    } = this.props;

    return (
      <Switch>
        <Route path={`${match.url}/playlists`} component={PlaylistsPage} />
        <Route
          path={match.url}
          component={() => {
            if (error) {
              return <Error>{error.message}</Error>;
            }

            if (isLoading) {
              return <Loader />;
            }

            return (
              <div>
                <Helmet>
                  <title>{category.name}</title>
                </Helmet>
                <Category {...category} />
              </div>
            );
          }}
        />
      </Switch>
    );
  }
}

const mapStateToProps = state => ({
  ...state,
  category: state.browse.category,
  error: state.browse.error,
  isLoading: state.browse.isLoading,
});

const mapDispatchToProps = dispatch => bindActionCreators(
  {
    fetchCategory,
  },
  dispatch,
);

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(CategoryPage);
