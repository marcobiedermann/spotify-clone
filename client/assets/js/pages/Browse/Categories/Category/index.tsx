import React, { FC } from 'react';
import { Helmet } from 'react-helmet';
import { RouteChildrenProps, useParams } from 'react-router-dom';
import useSWR from 'swr';
import Category from '../../../../components/Category';
import Error from '../../../../components/Error';
import Loader from '../../../../components/Loader';

interface Params {
  categoryId: string;
}

const CategoryPage: FC<RouteChildrenProps> = () => {
  const { categoryId } = useParams<Params>();
  const { data, error } = useSWR(`/v1/browse/categories/${categoryId}`);

  if (error) {
    return <Error>{error.message}</Error>;
  }

  if (!data) {
    return <Loader />;
  }

  const { name } = data;

  return (
    <>
      <Helmet>
        <title>{name}</title>
      </Helmet>
      <Category {...data} />
    </>
  );
};

export default CategoryPage;
