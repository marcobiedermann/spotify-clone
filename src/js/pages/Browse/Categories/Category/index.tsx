import React from 'react';
import { Helmet } from 'react-helmet';
import { useParams } from 'react-router-dom';
import { Category, Error, Loader } from '../../../../components';
import { useBrowseCategory } from '../../../../hooks/browse/categories';

function CategoryPage(): JSX.Element {
  const { categoryId } = useParams();
  const { data, error, isError, isLoading } = useBrowseCategory(categoryId!);

  if (isError) {
    return <Error>{error.message}</Error>;
  }

  if (isLoading) {
    return <Loader />;
  }

  const { name } = data;

  return (
    <>
      <Helmet>
        <title>{name}</title>
      </Helmet>
      <Category {...data} />
    </>
  );
}

export default CategoryPage;
