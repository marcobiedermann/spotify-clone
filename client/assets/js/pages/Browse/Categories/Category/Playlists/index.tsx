import React, { FC } from 'react';
import { Helmet } from 'react-helmet';
import { RouteChildrenProps, useParams } from 'react-router-dom';
import useSWR from 'swr';
import Error from '../../../../../components/Error';
import Loader from '../../../../../components/Loader';
import Playlists from '../../../../../components/Playlists';

interface Params {
  categoryId: string;
}

const PlaylistsPage: FC<RouteChildrenProps> = () => {
  const { categoryId } = useParams<Params>();
  const { data, error } = useSWR(`/v1/browse/categories/${categoryId}/playlists`);

  if (error) {
    return <Error>{error.message}</Error>;
  }

  if (!data) {
    return <Loader />;
  }

  const { playlists } = data;

  return (
    <>
      <Helmet>
        <title>Playlists</title>
      </Helmet>
      <Playlists items={playlists.items} />
    </>
  );
};

export default PlaylistsPage;
