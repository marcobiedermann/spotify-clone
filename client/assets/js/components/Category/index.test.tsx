import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import Category from '.';
import categoryFixture from './__fixtures__';

it('renders correctly', () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <Category {...categoryFixture} />
      </MemoryRouter>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
