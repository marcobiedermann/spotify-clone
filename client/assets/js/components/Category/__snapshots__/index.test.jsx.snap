// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<figure
  className="category"
>
  <figcaption>
    <a
      href="/browse/categories//playlists"
      onClick={[Function]}
    >
      
    </a>
  </figcaption>
</figure>
`;
